Performance enhancement of GitTopLevelDir in shell pkg

RaphSku opened the pull request:
## Description

Fixes #2344.

- GitTopLevelDir will only run `git rev-parse --show-toplevel` at the beginning and then re-use the previously found git root path
- A simple benchmark test is added in order to compare the old code vs. the new code in terms of performance

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [X] Update the docs.
I didn't update the docs since I am not changing any function signatures or behavior, I am just making the specified code more efficient. If I missed some document where my change chould be described or added to, please tell me.

- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

GitTopLevelDir will only run `git rev-parse --show-toplevel` at the beginning and then re-use the previously found git root path
Added / Removed / Updated [X].

## Gist
Gist: https://gist.github.com/RaphSku/8cb9aae84ea374de85d621b6cc6bf8f5
for the test output and benchmark

denis256 commented on shell/run_shell_cmd_test.go:
```diff
@@ -53,3 +53,13 @@ func TestRunShellOutputToStderrAndStdout(t *testing.T) {
 	assert.True(t, len(stdout.String()) == 0, "No output to stdout")
 }
+
+func BenchmarkPerformanceOfGitTopLevelDir(b *testing.B) {
```
Hi, will be helpful to have a functional test that will validate cache usage

RaphSku replied:
Agree, I'll add it

RaphSku replied:
Okay, I have added a test 👍

denis256 commented on shell/run_shell_cmd.go:
```diff
@@ -230,23 +230,29 @@ type CmdOutput struct {
 	Stderr string
 }
 
+var gitTopLevelDir string = ""
```
Not sure if this approach will work in case of multiple git repositories

RaphSku replied:
Good point, I'll consider this

RaphSku commented:
@denis256 Okay, so I have added a test for the caching. Previously, the StringCache, etc. were all in the config package. The problem with that is, that I cannot use them in the shell package due to circular import. So I have moved out all of the caches except the `TerragruntConfigCache` into a separate package. 

Also, I have made the cache structs generic with type constraint since the implementation of the `StringCache` and `options.IAMRoleOptions` are practically identical.

joaocc commented:
Hi. Any updates on merging this PR? Thanks

denis256 commented:
Hi,
should be fixed conflicts on files:
```
Conflicting files
config/config.go
```

RaphSku commented:
@denis256 I've updated the branch and fixed the conflict

denis256 commented on cache/cache.go:
```diff
@@ -0,0 +1,49 @@
+	Mutex *sync.Mutex
+}
+
+// NewStringCache - create new generic cache
```
I think should be `NewGenericCache`, to match function name...

RaphSku replied:
Thanks for the correction

denis256 commented on shell/run_shell_cmd.go:
```diff
@@ -268,23 +269,31 @@ type CmdOutput struct {
-	if err != nil {
-		return "", err
+	gitTopLevelDir, found := gitTopLevelDirs.Get(path)
+	if !found {
```
Nested blocks can be reduced if execution flow will exit if value is cached
```
	if gitTopLevelDir, found := gitTopLevelDirs.Load(path); found {
		return gitTopLevelDir.(string), nil
	}

...
```

RaphSku replied:
Yes, that is indeed more elegant, I corrected it, thanks

denis256 commented on shell/run_shell_cmd_test.go:
```diff
@@ -53,3 +53,36 @@ func TestRunShellOutputToStderrAndStdout(t *testing.T) {
+	expectedResult := strings.TrimSpace(cmd.Stdout)
+
+	actualResult, err := GitTopLevelDir(terragruntOptions, path)
+	assert.Nil(t, err, "Unexpected error executing GitTopLevelDir: %v", err)
```
I was wondering why can't be used `assert.NoError()`

RaphSku replied:
Exchanged

denis256 commented on shell/run_shell_cmd_test.go:
```diff
@@ -53,3 +53,36 @@ func TestRunShellOutputToStderrAndStdout(t *testing.T) {
+func BenchmarkPerformanceOfGitTopLevelDir(b *testing.B) {
+	for i := 0; i <= b.N; i++ {
+		terragruntOptions, err := options.NewTerragruntOptionsForTest("")
+		assert.Nil(b, err, "Unexpected error creating NewTerragruntOptionsForTest: %v", err)
```
`assert.NoError()`

denis256 commented:
This PR still misses integration tests(`test` directory) which will validate that `git rev-parse` is executed only once and cached value is returned

alikhil commented:
@RaphSku hello! How are you doing? 

Do you need assistance to finish your PR?

RaphSku commented:
> @RaphSku hello! How are you doing?
> 
> Do you need assistance to finish your PR?

Hi Alik, it would be great if you could assist me! Unfortunately, I didn't have much time on my hands. Thank you for asking.

denis256 commented:
Included in https://github.com/gruntwork-io/terragrunt/pull/3307
